internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,225 +6,655 @@
+	)
+
+	rec.shutdown = func() {
+		cleanup.Stop()
+		rec.activeUpstreams.Stop()
+		rec.activeUpstreams.DeleteAll()
+		waitForEvicts()
+		<-stopped
+	}
+
+	return rec
+}
+
+// Reconcile reconciles the workload proxy load balancers for a cluster. If the ReconcileData is nil or doesn't contain
+// any aliases, the cluster will be removed from the load balancer.
+func (rec *Reconciler) Reconcile(clusterID resource.ID, rd *ReconcileData) error {
+	logger := rec.logger.With(zap.String("cluster", clusterID), zap.Strings("upstreams", rd.GetHosts()))
 
-	lb, lbSts.lb = lbSts.lb, nil
+	logger.Info("reconcile LBs")
+
+	rec.mu.Lock()
+	defer rec.mu.Unlock()
+
+	rec.aliasToCluster.ReplaceCluster(clusterID, rd, func(alias) port {
+		got := rec.activeUpstreams.Get(clusterID)
+		if got == nil {
+			return ""
+		}
+
+		h, p := rec.aliasToCluster.ActiveHostPort(clusterID)
+		if p == "" {
+			rec.activeUpstreams.Delete(clusterID)
+
+			return ""
+		}
+
+		listLogger := rec.logger.With(
+			zap.String("cluster", clusterID),
+			zap.Strings("upstreams", rd.GetHosts()),
+			zap.Stringer("probe_port", p),
+		)
+
+		got.Value().Reconcile(asRemotes(h, p, listLogger))
+
+		return p
+	})
+
+	if rd == nil || len(rd.AliasPort) == 0 {
+		rec.activeUpstreams.Delete(clusterID)
+
+		rec.logger.Info("removed cluster from LB")
 
-	if lb == nil {
 		return nil
 	}
 
-	if err := lb.Close(); err != nil {
-		return fmt.Errorf("failed to close LB: %w", err)
+	aliases := slices.Collect(rd.AliasesData())
+	ports := xslices.Map(aliases, rd.PortForAlias)
+
+	logger = logger.With(zap.Strings("aliases", aliases), zap.Strings("ports", ports))
+
+	logger.Info("reconciled cluster in LB")
+
+	h, p := rec.aliasToCluster.ActiveHostPort(clusterID)
+	if p == "" {
+		rec.activeUpstreams.Delete(clusterID)
+
+		return nil
 	}
 
-	if err := lb.Wait(); err != nil && !errors.Is(err, net.ErrClosed) {
-		return fmt.Errorf("failed to wait for load balancer: %w", err)
+	listLogger := rec.logger.With(
+		zap.String("cluster", clusterID),
+		zap.Strings("upstreams", rd.GetHosts()),
+		zap.Stringer("probe_port", p),
+	)
+
+	got := rec.activeUpstreams.Get(clusterID)
+	if got == nil {
+		l, listErr := upstream.NewListWithCmp(asRemotes(h, p, listLogger), (*remote).Equal)
+		if listErr != nil {
+			return fmt.Errorf("failed to create upstream list for cluster %q: %w", clusterID, listErr)
+		}
+
+		got = rec.activeUpstreams.Set(clusterID, l, ttlcache.DefaultTTL)
 	}
 
+	got.Value().Reconcile(asRemotes(h, p, listLogger))
+
 	return nil
 }
 
-// Reconciler reconciles the load balancers for a cluster.
-type Reconciler struct {
-	clusterToAliasToLBStatus map[resource.ID]map[string]*lbStatus
-	aliasToCluster           map[string]resource.ID
-
-	connProvider *memconn.Provider
-	logger       *zap.Logger
-	lbLogger     *zap.Logger
-	logLevel     zapcore.Level
-	mu           sync.Mutex
+// ReconcileData is the data structure used to reconcile the load balancer for a specific cluster.
+type ReconcileData struct {
```

Comment:
I would appreciate if these small pieces would be split into separate files/internal packages and some tests added for them to simply understand what is going on better, as this file contains at least 5 or 6 different things

Reply:
Okay, done! Added tests.